### PR TITLE
MB-6431 additional timing output for benchmarking too document viewer

### PR DESCRIPTION
**Pull request by donaldthai**
## Description

> Given that I am a MilMove Engineer
>  When I run the test to load the orders viewer
>  Then I see an output table showing me the cumulative time spent Downloading
>       and the cumulative time spent waiting for the server

> AC:
> 
> We are able to extract and output the following metrics:
> 
> - Time to interactive 
> 
> - Total blocking time
> 
> - Largest Contentful Paint 

## Reviewer Notes

Is there anything you would like reviewers to give additional scrutiny?
- The lighthouse metrics may be slightly off from the performance timing since it's using a simulated network throttle. I think we'll have to fine tune the throttling config more later.

## Setup

Add any steps or code to run in this section to help others prepare to run your code:

**Locally**
```sh
make db_bandwidth_up // to create documents
make server_run
make office_client_run
node puppeteer/benchmarking.js run -m network-comparison
```

**S3**
Uncomment from `.envrc` file:
```
export STORAGE_BACKEND=s3
```
Terminal
```sh
// reload configs
direnv allow

// upload documents to s3
aws-vault exec transcom-gov-dev -- make db_bandwidth_up

// key in your aws OTP
aws-vault exec transcom-gov-dev -- make server_run

// run client
make office_client_run

// run script
node puppeteer/benchmarking.js run -m network-comparison
```

## Code Review Verification Steps

* [ ] If the change is risky, it has been tested in experimental before merging.
* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#logging)
* [ ] The requirements listed in [Querying the Database Safely](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#querying-the-database-safely) have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](https://github.com/transcom/mymove/wiki/migrate-the-database#zero-downtime-migrations))
  * [ ] Have been communicated to #g-database
  * [ ] Secure migrations have been tested following the instructions in our [docs](https://github.com/transcom/mymove/wiki/migrate-the-database#secure-migrations)
* [ ] There are no aXe warnings for UI.
* [ ] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] User facing changes have been reviewed by design.
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-6431) for this change
* [Lighthouse audit description](https://web.dev/lighthouse-performance)

## Screenshots

image

### CODE REVIEW

**Review by Ryan-Koch — APPROVED** [review]

**Review by jenniferlynparsons — APPROVED** [review]
Screen Shot 2021-02-19 at 4 12 00 PM
 Looks like it's working as expected. Do we know why the `fast` option has some of the longest times except for `Performance timing`? I would think it would have the fasted `Total blocking time` but it has the longest.

**Comment by duncan-truss** [comment]
> Screen Shot 2021-02-19 at 4 12 00 PM
> Looks like it's working as expected. Do we know why the `fast` option has some of the longest times except for `Performance timing`? I would think it would have the fasted `Total blocking time` but it has the longest.

Yea it feels like the last 3 metrics are constant, is it waiting for some event before calculating those?  For fast what does it mean for performance timing to be less than the time to interactive?

**Comment by donaldthai** [comment]
> > Screen Shot 2021-02-19 at 4 12 00 PM
> > Looks like it's working as expected. Do we know why the `fast` option has some of the longest times except for `Performance timing`? I would think it would have the fasted `Total blocking time` but it has the longest.
> 
> Yea it feels like the last 3 metrics are constant, is it waiting for some event before calculating those? For fast what does it mean for performance timing to be less than the time to interactive?

It could be possible I fudged up the lighthouse throttle settings. 😅

**Comment by donaldthai** [comment]
@duncan-truss @jenniferlynparsons alright I just updated the throttling values. The values should be more accurate now.

image

**Review by jenniferlynparsons — APPROVED** [review]
Looks good. This is exciting!

File: puppeteer/scenarios.js [hunk 1]
```diff
@@ -119,9 +182,43 @@ const totalDuration = async (host, config, debug) => {
     }),
   );
 
+  const url = `${host}/moves/${locatorValue}/orders`;
+  const lhOptions = {
+    port: new URL(browser.wsEndpoint()).port,
+    logLevel: 'info',
```
Comment (duncan-truss) [inline puppeteer/scenarios.js:188]: Did you mean to turn this logging of Lighthouse off by default?  Maybe we can reuse the verbose flag to toggle this on?

**Review by duncan-truss — APPROVED** [review]
Nice work 🎉! I was able to save the reports and open them in the Performance devtools tab and the lighthouse report viewer.

File: puppeteer/scenarios.js [hunk 2]
```diff
@@ -66,12 +79,62 @@ const setupCPU = async (config, page) => {
...
-const totalDuration = async (host, config, debug) => {
+const lighthouseFromPuppeteer = async (url, options, config = null, saveReports = false, networkProfileName = '') => {
+  // Run Lighthouse
+  const { lhr, artifacts } = await lighthouse(url, options, config);
+
+  // For debugging
+  const json = reportGenerator.generateReport(lhr, 'json');
+  if (saveReports) {
+    const trace = reportGenerator.generateReport(artifacts, 'json');
+    const lhReportPath = `puppeteer/lighthouse-report-${networkProfileName}.json`;
```
Comment (duncan-truss) [inline puppeteer/scenarios.js:98]: As a TODO we should document how someone can view these reports in the Performance devtools tab and the Lighthouse viewer.

**Pull request merged**